extensions/amp-carousel/0.2/amp-carousel.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ActionTrust} from '../../../src/action-constants';
+import {CSS} from '../../../build/amp-carousel-0.2.css';
+import {Carousel} from './carousel.js';
+import {dev} from '../../../src/log';
+import {htmlFor} from '../../../src/static-template';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {toArray} from '../../../src/types';
+
+/**
+ * @param {!Element} el The Element to check.
+ * @return {boolean} Whether or not the Element is a sizer Element.
+ */
+function isSizer(el) {
+  return el.tagName == 'I-AMPHTML-SIZER';
+}
+
+class AmpCarousel extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {number} */
+    this.advanceCount_ = 1;
+
+    /** @private {?Carousel} */
+    this.carousel_ = null;
+
+    /** @private {!Array<!Element>} */
+    this.slides_ = [];
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    const {element, win} = this;
+    // Grab the slides up front so we can place them later.
+    this.slides_ = toArray(element.children).filter(c => !isSizer(c));
+    // Create the carousel's inner DOM.
+    element.appendChild(this.renderContainerDom_());
+
+    const scrollContainer = dev().assertElement(
+        this.element.querySelector('.i-amphtml-carousel-scroll'));
+
+    this.carousel_ = new Carousel({
+      win,
+      element,
+      scrollContainer,
+      runMutate: cb => this.mutateElement(cb),
```

Comment:
BTW, this is an interesting approach to abstracting our inner vsync layer while keeping internal conformance.

A couple of thoughts:

- Do you see `measure` as a necessary op in the future? 
- Would there be value to simplifying to a vsync interface containing the measure/mutate methods and/or something conceptually compatible with other frameworks?

Reply:
> BTW, this is an interesting approach to abstracting our inner vsync layer while keeping internal conformance.

The original idea was to make the internal carousel implementation as standalone from AMP as possible, to make it easy for others to use. I initially passed in the debounce function as well. One thing I found though is that some things are less easy to pass through all the way (e.g. `setStyles).

> A couple of thoughts:

> * Do you see measure as a necessary op in the future?

I initially had a `runMeasure`, but after speaking with @jridgewell, it seems like there isn't that much value to it. If you do a measure whenever, it should be fine since all the mutates are batched together.

> * Would there be value to simplifying to a vsync interface containing the measure/mutate methods and/or something conceptually compatible with other frameworks?

I think for this to work across frameworks and gain adoption of a single solution, it would ultimately need to be a browser standard. I team I was on previously did reach out to see if there was interest in a standard from the Chrome team, but they were hesitant at the time (~2 years ago).